allennlp/models/constituency_parser.py

```diff
@@ -167,6 +177,13 @@ def forward(self,  # type: ignore
                 metric(logits, span_labels, span_mask)
             output_dict["loss"] = loss
 
+        # The evalb score is expensive to compute, so we only compute
+        # it for the validation # and test sets.
+        if gold_trees is not None and self._evalb_score is not None and not self.training:
+            decoded_output = self.decode(output_dict)
```

Comment:
Oh yeah.... I was wondering about this from your comment in the previous PR which I ignored because I hadn't thought of this.

Reply:
Actually @matt-gardner this seems fine to me - when running predict, you don't have the gold trees, so actually you wouldn't call it twice. This was why I ignored your comment previously - I have actually made this just now because I hadn't had a cup of coffee, so if you prefer it I can switch.

Reply:
It still seems like it's misusing the API to do it this way - this isn't how `decode` was intended to be used.  I'd recommend switching it to match how we currently use these functions, and thinking about whether we can improve this whole `forward` / `decode` distinction.  I've wondered for a while if `decode` was the wrong name for that function, because you pretty much always want to do any real "decoding" in `forward`, so you can evaluate the result in validation.  That function is much more about converting integers to strings.

Reply:
I broadly agree, although now I basically have the same function which just also get's called in decode - i.e there is no actual execution difference. I changed it regardless